osu.Game.Rulesets.Osu/Mods/OsuModBubbles.cs

```diff
@@ -0,0 +1,261 @@
+using osu.Game.Scoring;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.Mods
+{
+    public partial class OsuModBubbles : ModWithVisibilityAdjustment, IApplicableToDrawableRuleset<OsuHitObject>, IApplicableToScoreProcessor
+    {
+        public override string Name => "Bubbles";
+
+        public override string Acronym => "BU";
+
+        public override LocalisableString Description => "Don't let their popping distract you!";
+
+        public override double ScoreMultiplier => 1;
+
+        public override ModType Type => ModType.Fun;
+
+        // Compatibility with these seems potentially feasible in the future, blocked for now because they don't work as one would expect
+        public override Type[] IncompatibleMods => new[] { typeof(OsuModBarrelRoll), typeof(OsuModMagnetised), typeof(OsuModRepel) };
+
+        private PlayfieldAdjustmentContainer adjustmentContainer = null!;
+        private BubbleContainer bubbleContainer = null!;
+
+        private readonly Bindable<int> currentCombo = new BindableInt();
+
+        private float maxSize;
+        private float bubbleRadius;
+        private double bubbleFade;
+
+        public ScoreRank AdjustRank(ScoreRank rank, double accuracy) => rank;
+
+        public void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
+        {
+            currentCombo.BindTo(scoreProcessor.Combo);
+            currentCombo.BindValueChanged(combo =>
+                maxSize = Math.Min(1.75f, (float)(1.25 + 0.005 * combo.NewValue)), true);
+        }
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
+        {
+            // Multiplying by 2 results in an initial size that is too large, hence 1.85 has been chosen
+            bubbleRadius = (float)(drawableRuleset.Beatmap.HitObjects.OfType<HitCircle>().First().Radius * 1.85f);
+            bubbleFade = drawableRuleset.Beatmap.HitObjects.OfType<HitCircle>().First().TimePreempt * 2;
+
+            // We want to hide the judgements since they are obscured by the BubbleDrawable (due to layering)
+            drawableRuleset.Playfield.DisplayJudgements.Value = false;
+
+            adjustmentContainer = drawableRuleset.CreatePlayfieldAdjustmentContainer();
+
+            adjustmentContainer.Add(bubbleContainer = new BubbleContainer());
+            drawableRuleset.KeyBindingInputManager.Add(adjustmentContainer);
+        }
+
+        protected override void ApplyIncreasedVisibilityState(DrawableHitObject hitObject, ArmedState state) => applyBubbleState(hitObject);
+
+        protected override void ApplyNormalVisibilityState(DrawableHitObject hitObject, ArmedState state) => applyBubbleState(hitObject);
+
+        private void applyBubbleState(DrawableHitObject drawableObject)
+        {
+            if (drawableObject is DrawableSlider slider)
+            {
+                slider.Body.OnSkinChanged += () => applySliderState(slider);
+                applySliderState(slider);
+            }
```

Comment:
What on earth is this??

Reply:
https://github.com/ppy/osu/blob/c078d3a3441e95f3c4ddfcf12d14626c6958a755/osu.Game.Rulesets.Osu/Mods/OsuModTraceable.cs#L54-L57 it's conceptually the same as this no? , eitherway not sure how it could be more optimised

Reply:
Well that one is probably broken too. But I'm still confused as to what it's doing *visually*.

Reply:
colouring the outer border of the slider, regardless of skin, might not have seen it since i assume you are using argon

Reply:
Why is a "bubble" mod doing that? Is it documented somewhere as to why this is the case?

Reply:
> Why is a "bubble" mod doing that? Is it documented somewhere as to why this is the case?

nothing beyond aesthetics,  i felt pretty much everything being coloured except slider edges clashed, can be removed.

Reply:
Inline comment any aesthetic choices, please.